File: Stack&Queue/TruckRunningThroughBridge.java
import java.util.Queue;
import java.util.LinkedList;

class Queue<T> {
	
	class Node<T> {
		private T data;
		private Node<T> next;
		
		public Node( T data ) {
			this.data = data;
		}
	}
	
	private Node<T> first;
	private Node<T> last;
	
	public void enQueue( T item ) {
		Node<T> t = new Node<T>( item );
		if ( last != null ) {
			last.next = t;
		}
		last = t;
		if ( first == null ) {
			first = last;
		}
	}
	
	public T remove() {
		if ( first == null ) {
			return null;
		}
		T data = first.data;
		first = first.next;
		if ( first == null ) {
			last = null;
		}
		return data;
	}
	
	public T peek() {
		if ( first == null ) {
			return null;
		}
		return first.data;
	}
	
	public boolean isEmpty() {
		return first == null;
	}
}

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        Queue< Integer > q = new LinkedList< Integer >();
        int cnt = 0;
        while( cnt < truck_weights.length ) {
            q.push( truck_weights[ cnt ] );
            ++cnt;
        }
        return answer;
    }
}

File: Stack&Queue/TruckRunningThroughBridge.cpp

#include <queue>
using namespace std;


class Truck
{
public:
	int weight;
};

int solution(int bridge_length, int weight, vector<int> truck_weights) {
	int answer = 0;
	queue<int> bridge;
	for (int i = 0; i < bridge_length; ++i) {
		bridge.push(0);
	}
	int load_weight = 0;
	int cnt = 0;
	vector<int>::iterator it = truck_weights.begin();
	while ( bridge.size() != 0 ) {
		cnt++;
		load_weight -= bridge.front();
		bridge.pop();
		
		if (it != truck_weights.end()) {
			if ((load_weight + *it) <= weight) {
				bridge.push(*it);
				load_weight += *it;
				++it;
			}
			else
			{
				bridge.push(0);
			}
		}
	}
	answer = cnt;
	return answer;
}